express-notes-api/server.js

```diff
@@ -0,0 +1,155 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+const app = express();
+app.use(express.json());
+
+// errors array
+const errorArr = [
+  { error: 'Id must be a positive integer.' },
+  { error: 'Cannot find note with id.' },
+  { error: 'Content cannot be empty.' },
+  { error: 'An unexpected error occurred.' },
+];
+
+async function parse() {
+  const dataJson = await readFile('data.json', 'utf8');
+  const data = JSON.parse(dataJson);
+  return data;
+}
+
+// read all entries function and route
+async function readAll() {
+  const data = await parse();
+  const notes = data.notes;
+  const notesArr = [];
+  for (const key in notes) {
+    notesArr.push(notes[key]);
+  }
+  return notesArr;
+}
+
+app.get('/api/notes', async (req, res) => {
+  const allNotes = await readAll();
+  res.status(200).json(allNotes);
+});
+
+// read by ID function and
+async function readOne(id) {
+  const data = await parse();
+  const notes = data.notes;
+  return notes[id];
+}
+
+app.get('/api/notes/:id', async (req, res) => {
+  let { id } = req.params;
+  id = Number(id);
+  if (id < 0) {
+    res.status(400).json(errorArr[0]);
+  } else {
+    const oneNote = await readOne(id);
+    if (oneNote === undefined) {
+      errorArr[1].error = `Cannot find note with id ${id}.`;
+      res.status(404).json(errorArr[1]);
+    } else {
+      res.status(200).json(oneNote);
+    }
+  }
+});
+
+// update data.json
+async function saveIt(data) {
+  const jsonString = JSON.stringify(data, null, 2);
+  await writeFile('data.json', jsonString);
+}
+
+// add new note
+async function writeNewNote(contentObj) {
+  const data = await parse();
+  const notes = data.notes;
+  contentObj.id = data.nextId;
+  notes[data.nextId] = contentObj;
+  data.nextId++;
+  await saveIt(data);
+}
+
+app.post('/api/notes', async (req, res) => {
+  const contentObj = req.body;
+  if (!contentObj.content) {
+    res.status(400).json(errorArr[2]);
+  } else {
+    try {
+      await writeNewNote(contentObj);
+      const data = await parse();
+      res.status(201).json(data.notes[data.nextId - 1]);
```

Comment:
If `writeNewNote` returned the note that it just wrote, you could grab the return value in line 81 and return it here in line 83. Then you wouldn't need to re-grab `data` on line 82.